### ISSUE TITLE
Notebook starts on one port, hub polls another?

### ISSUE THREAD

**Issue by parente** [jupyterhub/jupyterhub#720]
I'm using hub 0.6.1 and sudospawner off the tip of master (0.3.0.dev0). Unless I'm misreading the logs, I'm running into a case where spawner is assigning the notebook server a port but is polling a different port for liveliness. The result is that the notebook server gets killed after 30 seconds.

For example, in the log below, the notebook server appears to be coming up on the port passed to it on the command line (37840), but hub is polling a totally different port and reporting the notebook never appeared on it (36013).

```
[I 160903 13:50:41 mediator:68] Spawning /opt/continuum/anaconda/envs/jupyterhub/bin/jupyterhub-singleuser 
--user=p-pparente --port=37840 --cookie-name=jupyter-hub-token-p-pparente 
--base-url=/user/p-pparente --hub-host= --hub-prefix=/hub/ 
--hub-api-url=http://127.0.0.1:8081/hub/api --ip=127.0.0.1 '--notebook-dir=~/notebooks'

[D 2016-09-03 13:50:42.086 JupyterHub spawner:316] Polling subprocess every 30s
[I 2016-09-03 13:50:42.717 p-pparente notebookapp:1128] Serving notebooks from local directory: /home/p-pparente/notebooks
[I 2016-09-03 13:50:42.717 p-pparente notebookapp:1128] 0 active kernels
[I 2016-09-03 13:50:42.717 p-pparente notebookapp:1128] The Jupyter Notebook is running at: http://127.0.0.1:37840/user/p-pparente/
[I 2016-09-03 13:50:42.717 p-pparente notebookapp:1129] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
...

[W 2016-09-03 13:51:12.165 JupyterHub user:264] p-pparente's server never showed up at http://127.0.0.1:36013/user/p-pparente after 30 seconds. Giving up

...

TimeoutError: Server at http://127.0.0.1:36013/user/p-pparente didn't respond in 30 seconds
```

**Comment by parente** [comment]
Bah. Found it: https://github.com/jupyterhub/sudospawner/commit/22b8b189af7d7b766d99f6be09f538541838acab

Another Lesson in Pinning My Dependencies!

tl;dr: The new sudospawner picks a random port and reports it back to the hub. Hub 0.6.1 expects the spawner to honor the assigned port.

**Comment by minrk** [comment]
@parente that commit doesn't change how ports are chosen. In both cases, `random_port()` is used to select the port, it's just a question of whether the value returned to the Hub goes in `self.port` or not. If you pin self.port to something with sudospawner, you can only have one user at a time on the entire Hub.

**Comment by parente** [comment]
Sorry I wasn't clear. I'm not pinning the port. I'm pinning my dependency on sudospawner 0.2.

**Comment by minrk** [comment]
Is your config somewhere? I'd like to understand how the bind and connect ports differ, because that seems to be a bug somewhere. I don't see how that sudospawner change would affect anything about how ports are identified to the Hub. I will poke around and see if there's something I misunderstood/forgot.

**Comment by minrk** [comment]
Oh, ha, that was quick. JupyterHub uses `self.port` to get the port in 0.7 (master), but `self.user.server.port` in 0.6, so the way I'm setting it in sudospawner won't take effect at the right time on 0.6. Should be fixed in jupyterhub/sudospawner#24
